spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java

```diff
@@ -40,37 +46,93 @@ final class WatcherUtil {
 	private WatcherUtil() {
 	}
 
-	static void onEvent(KubernetesObject kubernetesObject, String label, long refreshDelay,
+	static void onEvent(KubernetesObject kubernetesObject, String label, String annotationName, long refreshDelay,
 			ScheduledExecutorService executorService, String type,
-			Function<KubernetesObject, Mono<Void>> triggerRefresh) {
+			BiFunction<KubernetesObject, String, Mono<Void>> triggerRefresh) {
 
 		String name = kubernetesObject.getMetadata().getName();
 		boolean isSpringCloudKubernetes = isSpringCloudKubernetes(kubernetesObject, label);
 
 		if (isSpringCloudKubernetes) {
 
-			LOG.debug(() -> "Scheduling remote refresh event to be published for " + type + ": " + name
-					+ " to be published in " + refreshDelay + " milliseconds");
-			executorService.schedule(() -> {
-				try {
-					triggerRefresh.apply(kubernetesObject).subscribe();
-				}
-				catch (Throwable t) {
-					LOG.warn(t, "Error when refreshing ConfigMap " + name);
-				}
-			}, refreshDelay, TimeUnit.MILLISECONDS);
+			Set<String> apps = apps(kubernetesObject, annotationName);
+
+			if (!apps.isEmpty()) {
```

Comment:
the crox of the change is here: if the annotation that says we should restart multiple apps is present - try to get apps from it, and send events based on those names

Reply:
We could simplify this a bit if we add the name to the set if its empty